### ISSUE TITLE
Best practices exception handling

### ISSUE THREAD

**Issue by paulo-reichert** [GalateaRaj/fuse-starter-java#224]
Issue Description
--
Fuse should showcase best practices exception handling for REST web services. 

The guidelines are as follow:
- [ ] services and layers below those should throw appropriate exceptions. --> #346
  - [ ] built-in Java exceptions should be used when available and appropriate. 
  - [ ] custom exceptions should enhance the understanding of the error condition.
  - [ ] exceptions can be checked or unchecked depending on what is appropriate for the error condition*.
- [ ] the front-ends should handle exceptions correctly --> #347
  - [ ] the REST front-end implements the `RestExceptionHandler` to centralize the handling of exceptions
  - [ ] the responses from the REST front-end should have the correct HTTP return codes 
  - [ ] the REST front-end should include a descriptive message about the exception on the body of the response.
  - [ ] for JMS we need to handle exceptions and put the offending message in a dead-letter queue.
- [ ] Documentation needs to explain alternative error handling, such as returning HTTP codes 2xx and the exception details in body. --> #345
- [ ] EntityNotFound exception should be reviewed and if not necessary deleted and replaced with the built-in option in Spring Data. --> #268 

\* basic guidelines for exception throwing: --> #348
- Exceptions should not control the flow of execution of the application. They are for exceptional scenarios only.
- **Checked exceptions** should be thrown when the caller can be expected to handle it and recover.
- **Runtime exceptions** should indicate errors in the way the client code is using your API and it isn't clear whether it can recover. 
- Use checked exceptions sparingly as they force the client code to handle them. When in doubt throw runtime exceptions.

Design
--
[Description of the root cause for the issue and how to resolve it. Should be complete before Review is approved]

Documentation Changes
--
[Description of changes necessary to the Fuse documentation due to the changes in response to this issue. Should be complete before Review is approved]

Test Evidence
--
[include relevant tests and output. Should be complete before Review is approved]

Validation in Develop
-- 
[Proof issue is resolved in Develop. Should be complete before Issue is closed]

**Comment by cbaltera** [comment]
Additional items to cover:
- inclusion of error codes
- what fields to include in custom exceptions (message, cause when relevant)
-- guidelines around what to include in message (what happened, what are you doing now)
- if/when it's okay to not expose the stack trace (avoid sending to external consumers, try to avoid logging same stack trace multiple times, etc.)

**Comment by cbaltera** [comment]
Remaining decision: how do we want to show all these things?  Heather points out that there aren't a lot of places to generate exceptions in the current state of the code.

**Comment by mikegajda** [comment]
Decision: Make JMS item a separate ticket and figure out how may Galatea applications actually use JMS.

Decision: split out description into separate tickets.

**Comment by cbaltera** [comment]
This has been split into separate tickets described above.  Closing.
